{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "from transformers import AutoTokenizer, EsmModel\n",
    "import torch\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_appris(unique_transcripts=True):\n",
    "    # generate doc string\n",
    "    \"\"\"\n",
    "    Load the appris data\n",
    "    :param unique_transcripts: whether to load only unique transcripts\n",
    "    :return: the appris data\n",
    "    \"\"\"\n",
    "    # ## load human appris\n",
    "    dir = '/h/phil/Documents/01_projects/contrastive_rna_representation/'\n",
    "\n",
    "    app_h = pd.read_csv(f'{dir}/data/appris_data_human.principal.txt', sep='\\t')\n",
    "    print(app_h['Gene ID'].duplicated().sum())\n",
    "    app_h['numeric_value'] = app_h['APPRIS Annotation'].str.split(':').str[1]\n",
    "    app_h['key_value'] = app_h['APPRIS Annotation'].str.split(':').str[0]\n",
    "    app_h = app_h.sort_values(\n",
    "        ['Gene ID', 'key_value', 'numeric_value', \"Transcript ID\"],\n",
    "        ascending=[True, False, True, True],\n",
    "    )\n",
    "    if unique_transcripts:\n",
    "        app_h = app_h[~app_h.duplicated('Gene ID')]\n",
    "        app_h = app_h[~app_h.duplicated('Gene name')]\n",
    "    return app_h\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tokenizer = AutoTokenizer.from_pretrained(\"facebook/esm2_t6_8M_UR50D\")\n",
    "\n",
    "model = EsmModel.from_pretrained(\"facebook/esm2_t6_8M_UR50D\")\n",
    "\n",
    "inputs = tokenizer(\"Hello, my dog is cute\", return_tensors=\"pt\")\n",
    "\n",
    "outputs = model(**inputs)\n",
    "\n",
    "last_hidden_states = outputs.last_hidden_state"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from genome_kit import Genome, Interval\n",
    "genome = Genome(\"gencode.v29\")\n",
    "my_genes=[\n",
    "    'PTGDR2', 'MERTK', 'CD46', 'ITGB3', 'TNFSF4', 'MSR1', 'ICOSLG', 'PROM1', 'FASLG',\n",
    "    'NT5E', 'GYPA', 'CD72', 'CCR4', 'KLRB1', 'CD2', 'PDCD1LG2', 'CD19', 'SDC1', \n",
    "    'KIR3DL1', 'CR1', 'CD63', 'FCRL5', 'DPP4', 'CD28', 'CD209', 'LY75', 'CD177',\n",
    "    'LAIR1', 'ITGA2B', 'CD200', 'ENTPD1', 'CD70', 'OLR1', 'CD27', 'LILRA4', 'CCR3',\n",
    "    'SELP', 'MS4A1', 'NOTCH1', 'IFNGR1', 'CD34', 'FAS', 'ERBB2', 'THBD', 'TNFRSF18',\n",
    "    'CD244', 'PDGFRA', 'TNFRSF14', 'CXCR5', 'PROCR', 'ITGAE', 'CDH5', 'CD9', 'CD44',\n",
    "    'TNFSF10', 'NRP1', 'IL2RA', 'ANPEP', 'CD99', 'TNFRSF17', 'ENPP3', 'ITGA2', 'IL2RB',\n",
    "    'PDGFRB', 'PDCD1', 'SIGLEC8', 'VCAM1', 'ITGA1', 'ICAM2', 'FCGR1A', 'CD1D', 'CD36',\n",
    "    'ICAM1', 'NECTIN2', 'MPL', 'ABCB1', 'KIT', 'CDH17', 'CD8A', 'MRC1', 'ARHGEF5', 'TEK',\n",
    "    'CD81', 'NCAM1', 'LRPPRC', 'CD93', 'LAMP1', 'CLEC10A', 'KLRK1', 'B3GAT1', 'TNFRSF4', \n",
    "    'CD55', 'ITGA7', 'CLEC12A', 'TFRC', 'FLT3', 'CD48', 'CD24', 'CRLF2', 'SELE', 'CD164', \n",
    "    'NGFR', 'FLT4', 'CD52', 'CXCR4', 'SLAMF7', 'FCRL4', 'CLEC9A', 'CLEC4C', 'IL4R', 'IL6R',\n",
    "    'NCR3', 'SPN', 'ICOS', 'CD86', 'PTPRC', 'NOTCH2', 'CR2', 'IL7R', 'CD4', 'BTLA', \n",
    "    'TREM1', 'CD207', 'CD1A', 'CDH2', 'SIRPA', 'ITGAX', 'XCR1', 'F3', 'ABCG2', 'LGALS9', \n",
    "    'SIGLEC1', 'CCR6', 'CD22', 'FCGR3B', 'C5AR2', 'CD47', 'LAG3', 'CCR5', 'L1CAM', \n",
    "    'TNFRSF13B', 'CD40LG', 'ITGAM', 'CD1C', 'KDR', 'ACKR2', 'CD96', 'TNFRSF8', 'CD83', \n",
    "    'CTLA4', 'CD80', 'ENG', 'ITGB1', 'TNFRSF13C', 'PECAM1', 'VTCN1', 'CD68', 'ITGB2',\n",
    "    'CX3CR1', 'CSF1R', 'IL3RA', 'TLR4', 'CD14', 'TIGIT', 'FUT4', 'CDH1', 'CEACAM8', \n",
    "    'CD38', 'TNFRSF9', 'CD226', 'CD79A', 'NCR1', 'CD163', 'CD109', 'CD274', 'ITGA4',\n",
    "    'CLEC1B', 'CD40', 'CCR2', 'GP1BA', 'CD79B', 'HAVCR2', 'MCAM', 'THY1', 'SLC7A5', \n",
    "    'CD69', 'CXCR6', 'PVR'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15791\n"
     ]
    }
   ],
   "source": [
    "app = load_appris()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 188/188 [00:05<00:00, 35.62it/s]\n"
     ]
    }
   ],
   "source": [
    "transcripts_dict = {}\n",
    "for gene in tqdm(my_genes):\n",
    "    transcript_name = app[app['Gene name'] == gene].iloc[0]['Transcript ID']\n",
    "    gene_object = [x for x in genome.genes if x.name == gene][0]\n",
    "    for transcript in gene_object.transcripts:\n",
    "        if transcript.id.split('.')[0] == transcript_name:\n",
    "            transcripts_dict[gene] = transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAY*\n"
     ]
    }
   ],
   "source": [
    "# DNA codon to amino acid lookup table\n",
    "dna_to_amino_acid_table = {\n",
    "    'TTT': 'F', 'TTC': 'F', 'TTA': 'L', 'TTG': 'L',\n",
    "    'CTT': 'L', 'CTC': 'L', 'CTA': 'L', 'CTG': 'L',\n",
    "    'ATT': 'I', 'ATC': 'I', 'ATA': 'I', 'ATG': 'M',\n",
    "    'GTT': 'V', 'GTC': 'V', 'GTA': 'V', 'GTG': 'V',\n",
    "    'TCT': 'S', 'TCC': 'S', 'TCA': 'S', 'TCG': 'S',\n",
    "    'CCT': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',\n",
    "    'ACT': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',\n",
    "    'GCT': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',\n",
    "    'TAT': 'Y', 'TAC': 'Y', 'TAA': '*', 'TAG': '*',\n",
    "    'CAT': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "    'AAT': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "    'GAT': 'D', 'GAC': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "    'TGT': 'C', 'TGC': 'C', 'TGA': '*', 'TGG': 'W',\n",
    "    'CGT': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',\n",
    "    'AGT': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "    'GGT': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'\n",
    "}\n",
    "\n",
    "# Function to translate a DNA sequence to an amino acid sequence\n",
    "def dna_sequence_to_amino_acids(dna_sequence):\n",
    "    # Ensure the sequence length is a multiple of 3\n",
    "    if len(dna_sequence) % 3 != 0:\n",
    "        return \"Error: DNA sequence length must be a multiple of 3.\"\n",
    "\n",
    "    amino_acid_sequence = \"\"\n",
    "    \n",
    "    # Iterate over the DNA sequence in steps of 3 nucleotides (1 codon)\n",
    "    for i in range(0, len(dna_sequence), 3):\n",
    "        codon = dna_sequence[i:i+3].upper()\n",
    "        amino_acid = dna_to_amino_acid_table.get(codon, \"Invalid\")\n",
    "        \n",
    "        if amino_acid == \"Invalid\":\n",
    "            return f\"Error: Invalid codon '{codon}' encountered.\"\n",
    "        \n",
    "        amino_acid_sequence += amino_acid\n",
    "        \n",
    "        # Stop translation if a stop codon is encountered\n",
    "        if amino_acid == \"*\":\n",
    "            break\n",
    "    \n",
    "    return amino_acid_sequence\n",
    "\n",
    "# Example usage:\n",
    "dna_sequence = \"ATGGCTTACTAG\"\n",
    "amino_acids = dna_sequence_to_amino_acids(dna_sequence)\n",
    "print(amino_acids)  # Output: MAY* (Methionine, Alanine, Tyrosine, Stop)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MSANATLKPLCPILEQMSRLQSHSNTSIRYIDHAAVLLHGLASLLGLVENGVILFVVGCRMRQTVVTTWVLHLALSDLLASASLPFFTYFLAVGHSWELGTTFCKLHSSIFFLNMFASGFLLSAISLDRCLQVVRPVWAQNHRTVAAAHKVCLVLWALAVLNTVPYFVFRDTISRLDGRIMCYYNVLLLNPGPDRDATCNSRQVALAVSKFLLAFLVPLAIIASSHAAVSLRLQHRGRRRPGRFVRLVAAVVAAFALCWGPYHVFSLLEARAHANPGLRPLVWRGLPFVTSLAFFNSVANPVLYVLTCPDMLRKLRRSLRTVLESVLVDDSELGGAGSSRRRRTSSTARSASPLALCSRPEEPRGPARLLGWLLGSCAASPQTGPLNRALSSTSS*'"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_transcript_amino_acid_seq(transcript, genome):\n",
    "    exons = transcript.cdss\n",
    "    seq = \"\"\n",
    "    for exon in exons:\n",
    "        seq += genome.dna(exon)\n",
    "    return dna_sequence_to_amino_acids(seq)\n",
    "\n",
    "get_transcript_amino_acid_seq(transcripts_dict['PTGDR2'], genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/186 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 186/186 [01:42<00:00,  1.82it/s]\n"
     ]
    }
   ],
   "source": [
    "esm_data = []\n",
    "for gene, transcript in tqdm(transcripts_dict.items()):\n",
    "    aa_seq = get_transcript_amino_acid_seq(transcript, genome)\n",
    "    # * assert appears once in aa_seq\n",
    "    assert aa_seq.count(\"*\") == 1\n",
    "    inputs = tokenizer(aa_seq, return_tensors=\"pt\")\n",
    "    output = model(**inputs)\n",
    "    output = output['pooler_output'].detach().numpy()\n",
    "    row = {\n",
    "        'gene': gene,\n",
    "        'transcript': transcript.id,\n",
    "    }\n",
    "    for i in range(output.shape[1]):\n",
    "        row[f'esm_{i}'] = output[0, i]\n",
    "    esm_data.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(esm_data).to_csv('../data/esm_data.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "      <th>esm_0</th>\n",
       "      <th>esm_1</th>\n",
       "      <th>esm_2</th>\n",
       "      <th>esm_3</th>\n",
       "      <th>esm_4</th>\n",
       "      <th>esm_5</th>\n",
       "      <th>esm_6</th>\n",
       "      <th>esm_7</th>\n",
       "      <th>...</th>\n",
       "      <th>esm_310</th>\n",
       "      <th>esm_311</th>\n",
       "      <th>esm_312</th>\n",
       "      <th>esm_313</th>\n",
       "      <th>esm_314</th>\n",
       "      <th>esm_315</th>\n",
       "      <th>esm_316</th>\n",
       "      <th>esm_317</th>\n",
       "      <th>esm_318</th>\n",
       "      <th>esm_319</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PTGDR2</td>\n",
       "      <td>ENST00000332539.4</td>\n",
       "      <td>0.194653</td>\n",
       "      <td>0.126780</td>\n",
       "      <td>0.182600</td>\n",
       "      <td>-0.029803</td>\n",
       "      <td>0.051965</td>\n",
       "      <td>0.026792</td>\n",
       "      <td>0.109676</td>\n",
       "      <td>-0.213097</td>\n",
       "      <td>...</td>\n",
       "      <td>0.128914</td>\n",
       "      <td>0.092471</td>\n",
       "      <td>0.115099</td>\n",
       "      <td>0.152762</td>\n",
       "      <td>0.111469</td>\n",
       "      <td>0.124649</td>\n",
       "      <td>0.054852</td>\n",
       "      <td>0.122809</td>\n",
       "      <td>-0.076046</td>\n",
       "      <td>-0.050964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MERTK</td>\n",
       "      <td>ENST00000295408.8</td>\n",
       "      <td>0.174655</td>\n",
       "      <td>0.162040</td>\n",
       "      <td>0.159292</td>\n",
       "      <td>-0.064161</td>\n",
       "      <td>0.037378</td>\n",
       "      <td>0.025869</td>\n",
       "      <td>0.214430</td>\n",
       "      <td>-0.168868</td>\n",
       "      <td>...</td>\n",
       "      <td>0.130357</td>\n",
       "      <td>0.049052</td>\n",
       "      <td>0.103070</td>\n",
       "      <td>0.135618</td>\n",
       "      <td>0.163095</td>\n",
       "      <td>-0.009734</td>\n",
       "      <td>0.029290</td>\n",
       "      <td>0.162163</td>\n",
       "      <td>-0.071493</td>\n",
       "      <td>0.131510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CD46</td>\n",
       "      <td>ENST00000358170.6</td>\n",
       "      <td>0.160673</td>\n",
       "      <td>0.180229</td>\n",
       "      <td>0.124241</td>\n",
       "      <td>-0.047515</td>\n",
       "      <td>0.121338</td>\n",
       "      <td>0.048503</td>\n",
       "      <td>0.204651</td>\n",
       "      <td>-0.230932</td>\n",
       "      <td>...</td>\n",
       "      <td>0.178895</td>\n",
       "      <td>0.039712</td>\n",
       "      <td>0.099157</td>\n",
       "      <td>-0.035398</td>\n",
       "      <td>0.227750</td>\n",
       "      <td>-0.011657</td>\n",
       "      <td>0.019348</td>\n",
       "      <td>0.188651</td>\n",
       "      <td>-0.027152</td>\n",
       "      <td>0.146914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ITGB3</td>\n",
       "      <td>ENST00000559488.5</td>\n",
       "      <td>0.033419</td>\n",
       "      <td>0.148301</td>\n",
       "      <td>0.171840</td>\n",
       "      <td>-0.122247</td>\n",
       "      <td>0.052923</td>\n",
       "      <td>0.050449</td>\n",
       "      <td>0.140216</td>\n",
       "      <td>-0.211933</td>\n",
       "      <td>...</td>\n",
       "      <td>0.101055</td>\n",
       "      <td>0.028358</td>\n",
       "      <td>0.171624</td>\n",
       "      <td>0.008692</td>\n",
       "      <td>0.149858</td>\n",
       "      <td>-0.025355</td>\n",
       "      <td>0.080045</td>\n",
       "      <td>0.208778</td>\n",
       "      <td>-0.035152</td>\n",
       "      <td>0.119216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TNFSF4</td>\n",
       "      <td>ENST00000281834.3</td>\n",
       "      <td>0.159453</td>\n",
       "      <td>0.045351</td>\n",
       "      <td>0.133093</td>\n",
       "      <td>-0.065311</td>\n",
       "      <td>-0.001200</td>\n",
       "      <td>0.090423</td>\n",
       "      <td>0.113904</td>\n",
       "      <td>-0.243310</td>\n",
       "      <td>...</td>\n",
       "      <td>0.225913</td>\n",
       "      <td>0.040134</td>\n",
       "      <td>0.083178</td>\n",
       "      <td>-0.037448</td>\n",
       "      <td>0.193334</td>\n",
       "      <td>-0.017796</td>\n",
       "      <td>0.067195</td>\n",
       "      <td>0.200707</td>\n",
       "      <td>0.061961</td>\n",
       "      <td>0.077091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>THY1</td>\n",
       "      <td>ENST00000284240.9</td>\n",
       "      <td>0.108301</td>\n",
       "      <td>0.146954</td>\n",
       "      <td>0.171239</td>\n",
       "      <td>-0.001321</td>\n",
       "      <td>0.000676</td>\n",
       "      <td>-0.037172</td>\n",
       "      <td>0.190913</td>\n",
       "      <td>-0.116044</td>\n",
       "      <td>...</td>\n",
       "      <td>0.157813</td>\n",
       "      <td>0.002127</td>\n",
       "      <td>0.212422</td>\n",
       "      <td>0.027532</td>\n",
       "      <td>0.151119</td>\n",
       "      <td>0.039846</td>\n",
       "      <td>0.055663</td>\n",
       "      <td>0.189299</td>\n",
       "      <td>-0.071068</td>\n",
       "      <td>0.024650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>SLC7A5</td>\n",
       "      <td>ENST00000261622.4</td>\n",
       "      <td>0.110616</td>\n",
       "      <td>0.120736</td>\n",
       "      <td>0.166708</td>\n",
       "      <td>-0.035123</td>\n",
       "      <td>0.000360</td>\n",
       "      <td>0.071963</td>\n",
       "      <td>0.124000</td>\n",
       "      <td>-0.362562</td>\n",
       "      <td>...</td>\n",
       "      <td>0.121050</td>\n",
       "      <td>0.019785</td>\n",
       "      <td>0.111488</td>\n",
       "      <td>0.042031</td>\n",
       "      <td>0.230152</td>\n",
       "      <td>-0.065543</td>\n",
       "      <td>0.002094</td>\n",
       "      <td>0.153272</td>\n",
       "      <td>-0.042976</td>\n",
       "      <td>0.102418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>CD69</td>\n",
       "      <td>ENST00000228434.7</td>\n",
       "      <td>0.197017</td>\n",
       "      <td>0.137450</td>\n",
       "      <td>0.162856</td>\n",
       "      <td>0.019329</td>\n",
       "      <td>0.073427</td>\n",
       "      <td>0.057510</td>\n",
       "      <td>0.155821</td>\n",
       "      <td>-0.286048</td>\n",
       "      <td>...</td>\n",
       "      <td>0.216885</td>\n",
       "      <td>0.054472</td>\n",
       "      <td>0.062408</td>\n",
       "      <td>-0.029178</td>\n",
       "      <td>0.193458</td>\n",
       "      <td>0.010828</td>\n",
       "      <td>0.010842</td>\n",
       "      <td>0.188487</td>\n",
       "      <td>-0.011039</td>\n",
       "      <td>0.043911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>CXCR6</td>\n",
       "      <td>ENST00000304552.4</td>\n",
       "      <td>0.182692</td>\n",
       "      <td>0.106208</td>\n",
       "      <td>0.124159</td>\n",
       "      <td>-0.011814</td>\n",
       "      <td>0.007098</td>\n",
       "      <td>0.064117</td>\n",
       "      <td>0.131658</td>\n",
       "      <td>-0.237451</td>\n",
       "      <td>...</td>\n",
       "      <td>0.229905</td>\n",
       "      <td>0.046099</td>\n",
       "      <td>0.081270</td>\n",
       "      <td>0.091075</td>\n",
       "      <td>0.220422</td>\n",
       "      <td>-0.002151</td>\n",
       "      <td>0.039243</td>\n",
       "      <td>0.161981</td>\n",
       "      <td>-0.020390</td>\n",
       "      <td>0.000847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>PVR</td>\n",
       "      <td>ENST00000425690.7</td>\n",
       "      <td>0.116172</td>\n",
       "      <td>0.178937</td>\n",
       "      <td>0.189908</td>\n",
       "      <td>-0.103309</td>\n",
       "      <td>0.049461</td>\n",
       "      <td>0.000310</td>\n",
       "      <td>0.204055</td>\n",
       "      <td>-0.165687</td>\n",
       "      <td>...</td>\n",
       "      <td>0.114715</td>\n",
       "      <td>0.069304</td>\n",
       "      <td>0.158498</td>\n",
       "      <td>0.079845</td>\n",
       "      <td>0.161891</td>\n",
       "      <td>0.072528</td>\n",
       "      <td>0.053655</td>\n",
       "      <td>0.168270</td>\n",
       "      <td>-0.097275</td>\n",
       "      <td>0.047861</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>186 rows × 322 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene         transcript     esm_0     esm_1     esm_2     esm_3  \\\n",
       "0    PTGDR2  ENST00000332539.4  0.194653  0.126780  0.182600 -0.029803   \n",
       "1     MERTK  ENST00000295408.8  0.174655  0.162040  0.159292 -0.064161   \n",
       "2      CD46  ENST00000358170.6  0.160673  0.180229  0.124241 -0.047515   \n",
       "3     ITGB3  ENST00000559488.5  0.033419  0.148301  0.171840 -0.122247   \n",
       "4    TNFSF4  ENST00000281834.3  0.159453  0.045351  0.133093 -0.065311   \n",
       "..      ...                ...       ...       ...       ...       ...   \n",
       "181    THY1  ENST00000284240.9  0.108301  0.146954  0.171239 -0.001321   \n",
       "182  SLC7A5  ENST00000261622.4  0.110616  0.120736  0.166708 -0.035123   \n",
       "183    CD69  ENST00000228434.7  0.197017  0.137450  0.162856  0.019329   \n",
       "184   CXCR6  ENST00000304552.4  0.182692  0.106208  0.124159 -0.011814   \n",
       "185     PVR  ENST00000425690.7  0.116172  0.178937  0.189908 -0.103309   \n",
       "\n",
       "        esm_4     esm_5     esm_6     esm_7  ...   esm_310   esm_311  \\\n",
       "0    0.051965  0.026792  0.109676 -0.213097  ...  0.128914  0.092471   \n",
       "1    0.037378  0.025869  0.214430 -0.168868  ...  0.130357  0.049052   \n",
       "2    0.121338  0.048503  0.204651 -0.230932  ...  0.178895  0.039712   \n",
       "3    0.052923  0.050449  0.140216 -0.211933  ...  0.101055  0.028358   \n",
       "4   -0.001200  0.090423  0.113904 -0.243310  ...  0.225913  0.040134   \n",
       "..        ...       ...       ...       ...  ...       ...       ...   \n",
       "181  0.000676 -0.037172  0.190913 -0.116044  ...  0.157813  0.002127   \n",
       "182  0.000360  0.071963  0.124000 -0.362562  ...  0.121050  0.019785   \n",
       "183  0.073427  0.057510  0.155821 -0.286048  ...  0.216885  0.054472   \n",
       "184  0.007098  0.064117  0.131658 -0.237451  ...  0.229905  0.046099   \n",
       "185  0.049461  0.000310  0.204055 -0.165687  ...  0.114715  0.069304   \n",
       "\n",
       "      esm_312   esm_313   esm_314   esm_315   esm_316   esm_317   esm_318  \\\n",
       "0    0.115099  0.152762  0.111469  0.124649  0.054852  0.122809 -0.076046   \n",
       "1    0.103070  0.135618  0.163095 -0.009734  0.029290  0.162163 -0.071493   \n",
       "2    0.099157 -0.035398  0.227750 -0.011657  0.019348  0.188651 -0.027152   \n",
       "3    0.171624  0.008692  0.149858 -0.025355  0.080045  0.208778 -0.035152   \n",
       "4    0.083178 -0.037448  0.193334 -0.017796  0.067195  0.200707  0.061961   \n",
       "..        ...       ...       ...       ...       ...       ...       ...   \n",
       "181  0.212422  0.027532  0.151119  0.039846  0.055663  0.189299 -0.071068   \n",
       "182  0.111488  0.042031  0.230152 -0.065543  0.002094  0.153272 -0.042976   \n",
       "183  0.062408 -0.029178  0.193458  0.010828  0.010842  0.188487 -0.011039   \n",
       "184  0.081270  0.091075  0.220422 -0.002151  0.039243  0.161981 -0.020390   \n",
       "185  0.158498  0.079845  0.161891  0.072528  0.053655  0.168270 -0.097275   \n",
       "\n",
       "      esm_319  \n",
       "0   -0.050964  \n",
       "1    0.131510  \n",
       "2    0.146914  \n",
       "3    0.119216  \n",
       "4    0.077091  \n",
       "..        ...  \n",
       "181  0.024650  \n",
       "182  0.102418  \n",
       "183  0.043911  \n",
       "184  0.000847  \n",
       "185  0.047861  \n",
       "\n",
       "[186 rows x 322 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('../data/esm_data.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cite_seq",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
